### PR TITLE
"Continue" button work

### PR DESCRIPTION
**Pull request by ashfurrow**
This is work for [PURCHASE-136](https://artsyproduct.atlassian.net/browse/PURCHASE-136), paired with @sweir27 on this. The PR does a few things:

- The `presentModalViewController` function on `SwitchBoard` now returns a promise that resolves when the completion block of the Objective-C `dismissViewController:animated:completion` is executed.
- Hooks up the "Continue" button of the bid result component to dismiss the BidFlow modal and, optionally, to open the LAI in Eigen if appropriate.
- Updates the example app to present the `BidFlow` component modally, to more realistically present the component, which entailed a few other changes to the app.
- Adds test coverage to the `BidResult` component and factors out common logic into new `describe` blocks.

### CODE REVIEW

File: src/lib/Components/Bidding/Screens/BidResult.tsx [hunk 1]
```diff
@@ -74,7 +83,7 @@ export class BidResult extends React.Component<BidResultProps> {
                 }}
               />
             ) : (
-              <BidGhostButton text="Continue" onPress={() => null} />
+              <BidGhostButton text="Continue" onPress={this.exitBidFlow} />
```
Comment (ashfurrow) [inline src/lib/Components/Bidding/Screens/BidResult.tsx]: I was wondering if there were problems with this syntax, binding to `this` or anything?

File: Example/Emission/AppDelegate.m [hunk 2]
```diff
@@ -185,7 +185,11 @@ - (void)setupEmissionWithUserID:(NSString *)userID accessToken:(NSString *)acces
     viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                                                                     target:self
                                                                                                     action:@selector(dismissModalViewController)];
-    [fromViewController.navigationController presentViewController:navigationController animated:YES completion:nil];
+    if (fromViewController.isBeingDismissed || fromViewController.navigationController.isBeingDismissed) {
+      [self.navigationController presentViewController:navigationController animated:YES completion:nil];
+    } else {
+      [fromViewController.navigationController presentViewController:navigationController animated:YES completion:nil];
```
Comment (ashfurrow) [inline Example/Emission/AppDelegate.m]: This needs to be documented.

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,8 @@
 * Adds the ability to style the text in the inverted button with a spinner - yuki24
 * Now the `<ConfirmBid>` component shows a spinner while loading - yuki24
 * Now the `<Input>` component updates the border color when the parent component updates the `error` prop - yuki24
+* `dismissModalViewController` now returns a promise - ash&sarah
```
Comment (orta) [inline CHANGELOG.md]: Should we see native code for this? I don't see any

Comment (ashfurrow) [reply CHANGELOG.md]: Yup, a small change here: https://github.com/artsy/emission/pull/1065/files#diff-c389a83eac4316bce2520620bd807d14R42

Comment (orta) [reply CHANGELOG.md]: Ace, can you bump the [native-code-version](https://github.com/artsy/emission/blob/develop-1.5.x/package.json#L4) in the package.json - wonder what's up with the [danger rule](https://github.com/artsy/emission/blob/develop-1.5.x/dangerfile.ts#L184-L195)

Comment (ashfurrow) [reply CHANGELOG.md]: Fixed the rule here, next CI should fail: https://github.com/artsy/emission/pull/1065/commits/cb73d7e960b231beab6d00fdadb3eabdaed546ef

**Comment by ashfurrow** [comment]
Had to resolve some conflicts wit this comment: https://github.com/artsy/emission/commit/dffab0fe65720e7ecfe5af575736c7ae7bf6f23e I have a few questions I'm going to leave inline.

File: src/lib/Components/Bidding/Screens/BidResult.tsx [hunk 4]
```diff
@@ -16,7 +18,7 @@ import { Title } from "../Components/Title"
 
 import { BidResult_sale_artwork } from "__generated__/BidResult_sale_artwork.graphql"
 
-const SHOW_TIMER_STATUSES = ["WINNING", "OUTBID", "RESERVE_NOT_MET"]
+const SHOW_TIMER_STATUSES = ["WINNING", "OUTBID", "RESERVE_NOT_MET", "ERROR_BID_LOW"]
```
Comment (ashfurrow) [inline src/lib/Components/Bidding/Screens/BidResult.tsx]: @sepans This was removed [here](https://github.com/artsy/emission/pull/1054/files#diff-daac583405fb039e935fb2fec6983390L19) but I'm not sure why. It seems like we want to show the bid-again button when `ERROR_BID_LOW` is the status, right? I think I'm missing something.

File: Pod/Assets/Emission.js.meta [hunk 5]
```diff
@@ -1 +1 @@
-8���E��o�����ɳ
\ No newline at end of file
+8��x�w�!����-:Q<
```
Comment (yuki24) [inline Pod/Assets/Emission.js.meta]: 😳

File: src/lib/Components/Bidding/Screens/BidResult.tsx (hunk 4 above)
Comment (sweir27) [reply src/lib/Components/Bidding/Screens/BidResult.tsx]: I believe `ERROR_BID_LOW` was renamed to `OUTBID` (we still control this whole flow 😄)

File: src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx [hunk 6]
```diff
@@ -44,66 +55,125 @@ const bid = {
...
+    const messageDescriptionMd =
+      "Another bidder placed a higher max bid or the same max bid before you did.  \
+ Bid again to take the lead."
+
+    it("renders timer and error message", () => {
+      const bidResult = (
+        <BidResult
+          winning={false}
+          sale_artwork={saleArtwork}
+          status="ERROR_BID_LOW"
```
Comment (sepans) [inline src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx]: This also needs to be "OUTBID", right?

File: src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx [hunk 7]
```diff
@@ -44,66 +55,125 @@ const bid = {
...
+      const component = shallow(bidResult)
+      expect(component.find("TimeLeftToBidDisplay")).toHaveLength(1)
+    })
+
+    it("pops to root when bid-again button is pressed", () => {
+      const bidResult = renderer.create(
+        <BidResult
+          winning={false}
+          sale_artwork={saleArtwork}
+          status="ERROR_BID_LOW"
```
Comment (sepans) [inline src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx]: Also 👆

File: src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx [hunk 8]
```diff
@@ -44,66 +55,125 @@ const bid = {
...
+
+      expect(popToTop).toHaveBeenCalled()
+    })
   })
-  it("doesn't render timer when live bidding is started", () => {
-    jest.useFakeTimers()
-    const status = "LIVE_BIDDING_STARTED"
+
+  describe("live bidding has started", () => {
+    const status = "ERROR_LIVE_BIDDING_STARTED"
```
Comment (sepans) [inline src/lib/Components/Bidding/Screens/__tests__/BidResult-tests.tsx]: And this needs to be "LIVE_BIDDING_STARTED"

File: src/lib/Components/Bidding/Screens/BidResult.tsx (hunk 4 above)
Comment (sepans) [reply src/lib/Components/Bidding/Screens/BidResult.tsx]: Yes we no longer need "ERROR_BID_LOW"

File: src/lib/Components/Bidding/Screens/BidResult.tsx [hunk 9]
```diff
@@ -41,6 +43,13 @@ export class BidResult extends React.Component<BidResultProps> {
     this.props.navigator.popToTop()
   }
 
+  exitBidFlow = async () => {
+    await SwitchBoard.dismissModalViewController(this)
+    if (this.props.status === "ERROR_LIVE_BIDDING_STARTED") {
```
Comment (sepans) [inline src/lib/Components/Bidding/Screens/BidResult.tsx]: Also this -> "LIVE_BIDDING_STARTED"

**Comment by ashfurrow** [comment]
Okay thanks @sepans and everyone, I'll make those changes 👍

**Comment by ashfurrow** [comment]
@sepans Agreed, it feels like we should be able to use the Metaphysics schema stored in Emission to tie it together a little better.

**Comment by ashfurrow** [comment]
@sepans So the bid messages linked to in metaphysics are all _failure_ messages, so I had to add `"SUCCESS"`, but we _could_ do something like this:

```diff
diff --git a/src/lib/Components/Bidding/Screens/BidResult.tsx b/src/lib/Components/Bidding/Screens/BidResult.tsx
index f2168138..fafc5937 100644
--- a/src/lib/Components/Bidding/Screens/BidResult.tsx
+++ b/src/lib/Components/Bidding/Screens/BidResult.tsx
@@ -20,6 +20,15 @@ import { BidResult_sale_artwork } from "__generated__/BidResult_sale_artwork.gra
 
 const SHOW_TIMER_STATUSES = ["WINNING", "OUTBID", "RESERVE_NOT_MET"]
 
+type BidResultStatus =
+  | "SUCCESS"
+  | "OUTBID"
+  | "RESERVE_NOT_MET"
+  | "SALE_CLOSED"
+  | "LIVE_BIDDING_STARTED"
+  | "BIDDER_NOT_QUALIFIED"
+  | "ERROR"
+
 interface Bid {
   display: string
   cents: number
@@ -29,7 +38,7 @@ interface BidResultProps {
   sale_artwork: BidResult_sale_artwork
   bid: Bid
   winning: boolean
-  status: string
+  status: BidResultStatus
   message_header?: string
   message_description_md?: string
   suggested_next_bid?: Bid
```

And get TypeScript to verify the values for `status` are valid. What do you think?

**Comment by ashfurrow** [comment]
Danger is failing intermittently, but this should be good to merge 👍

**Pull request merged**